### PR TITLE
GraphQL error customization, coerce input/output fix, variable object coersion, improved Giraffe integration

### PR DESCRIPTION
**Pull request by xperiandri**
1. Implemented `IGQLError` and `IGQLErrorExtensions` to return errors according to the latest GraphQL specification
2. Added `InputParameterValue` to distinguish between `InlineConstant` which comes from GraphQL operation and `Variable` which comes from variables
3. Renamed `CoerceValue` to `CoerceOutput`
4. Implemented the whole object construction that comes from variables using coercion of each scalar that it requires
5. Made Giraffe handler more concise and traceable

### CODE REVIEW

**Comment by xperiandri** [comment]
2 tests are broken:
One because of the need to use `Skippable` instead of `Option`
The second is because of the change of the `Observable` order between home-brew implementation and `FSharp.Control.Reactive`

**Comment by xperiandri** [comment]
WebSockets middleware is broken
React samples not tested

File: src/FSharp.Data.GraphQL.Server/Executor.fs [hunk 1]
```diff
@@ -119,19 +115,20 @@ type Executor<'Root>(schema: ISchema<'Root>, middlewares : IExecutorMiddleware s
                           Errors = errors
                           FieldExecuteMap = fieldExecuteMap
                           Metadata = executionPlan.Metadata }
-                    let! res = runMiddlewares (fun x -> x.ExecuteOperationAsync) executionCtx executeOperation
+                    let! res = runMiddlewares (fun x -> x.ExecuteOperationAsync) executionCtx executeOperation |> AsyncVal.toAsync
                     return prepareOutput res
                 with
-                | ex -> return prepareOutput(GQLResponse.Error(ex.ToString(), executionPlan.Metadata))
+                | :? GraphQLException as ex -> return prepareOutput(GQLExecutionResult.Error(documentId, ex, executionPlan.Metadata))
+                | ex -> return prepareOutput(GQLExecutionResult.Error(documentId, ex.ToString(), executionPlan.Metadata)) // TODO: Handle better
```
Comment (valbers) [inline src/FSharp.Data.GraphQL.Server/Executor.fs]: Please see: https://github.com/fsprojects/FSharp.Data.GraphQL/issues/425 (I just created this issue before looking at this pull request here ^^).

Comment (xperiandri) [reply src/FSharp.Data.GraphQL.Server/Executor.fs]: Yes, the fix will come in that PR. So you will be able to do something like this.
```F#
// TODO: Handle your exceptions here and transform to IGQLError, optionally implement IGQLErrorExtensions for details
// GraphQLException already implements IGQLError
let parseError (ex : exn) =
    match ex with
    | :? ResolverErrorException as err ->
       [err :> IGQLError]
    | :? ResolverErrorExceptions as errs ->
      errs.InnerExceptions |> Seq.cast<IGQLError> |> Seq.toList
    | _  ->
        [{ new IGQLError with member _.Message = ex.GetType().FullName
            interface IGQLErrorExtensions with
            member error.Extensions =
                seq {
                    // TODO: Do not expose in production
                    kvp "stackTrace" (box ex.StackTrace)
                    match ex with
                    | :? AggregateException as agg ->
                       kvp "innerExceptions" (box agg.InnerExceptions)
                    | _ -> ()
                }
                |> ImmutableDictionary.CreateRange<string, obj> :> IReadOnlyDictionary<string, obj>
                |> ValueSome
            }]

let schemaConfig =
    { SchemaConfig.Default with ParseError = parseError }
```

Comment (valbers) [reply src/FSharp.Data.GraphQL.Server/Executor.fs]: great! :+1:

**Comment by xperiandri** [comment]
Updated branch

**Comment by xperiandri** [comment]
Added some fixes. But integration tests are fixed in #436

**Comment by valbers** [comment]
Do we want to finish https://github.com/fsprojects/FSharp.Data.GraphQL/pull/436 (PR from `input-objects-creation-with-coercion` to `IGQLError`) before approving this one (which is a PR from `IGQLError` to `dev`)?

**Comment by xperiandri** [comment]
Yes.
1. Merge #436 into #418
2. Merge #418 into dev
3. Rebase and merge #437 into dev
4. We are ready to apply your #430

**Review by valbers — APPROVED** [review]
LGTM

**Review by Juriyx — APPROVED** [review]

**Pull request merged**